### PR TITLE
[LW] Frontpage Tabs for the masses (including a recommendations one)

### PR DESCRIPTION
**Pull request by darkruby501**
LessWrong is introducing a new tabs UI to our front page (borrowing from the EA Forum TagBar picker element). This greatly reduces clutter of the many sections that had accumulated. 

This change is motivated by our experimentation with a recommendations-engine driver posts list where Latest Posts goes. Locally, we've been trying out Recombee recommendation-saas. 

The frontpage varies due to multiple factors:
* logged-in vs logged-out
* screen size
* How many tabs the user has

The set of possible tabs at present is:
- Latest (Posts) (classic algorithm)
- Recommendations (list combining approx HN algorithm and Recombee recommendations)
- Bookmarks (if the user has any)
- Continue Reading (if the user has any)

The Tabs UI is only show to logged-in users (because we haven't set up recommendations to work for logged out users yet)

So logged out users get:

**Logged-in, New User, Only Two Tabs**
LessWrong 2024-04-19 16-38-56

**Logged-in, All Tabs**
LessWrong 2024-04-19 16-40-16

*Frontpage Tag Filters Available by clicking gear icon to expand*
LessWrong 2024-04-19 16-41-56

**Logged-in, All Tabs** – scroll UI enabled on small screens
LessWrong 2024-04-19 16-41-07

**Logged-Out** – no tabs for now, tag filter settings always open on large screens, collapsed with settings button on small screens

LessWrong 2024-04-19 16-43-19

LessWrong 2024-04-19 16-43-30

The scroll UI ended up being very jank to handle the varying possibilities of having or not having enough tabs to warrant showing the scrolling UI/fade, and knowing what to render in SSR. Not sure if we'll solved this by just having more tags or something else. Currently bunch of breakpoints calculated by hand + trial and error.

----
Other code elements:
This moves LessWrong Frontpage logic into a new component `LWHomePosts` that lives inside LWHome instead of the shared `HomeLatestPosts` component.

### CODE REVIEW

File: packages/lesswrong/components/bookmarks/BookmarksPage.tsx [hunk 1]
```diff
@@ -65,9 +65,9 @@ const BookmarksPage = ({ classes }: {
 
   return <AnalyticsContext pageContext="bookmarksPage" capturePostItemOnMount>
     <SingleColumnSection>
-      <Typography variant="display2" className={classes.headline}>
+      {/* <Typography variant="display2" className={classes.headline}>
         {preferredHeadingCase(`Saved & Read`)}
-      </Typography>
+      </Typography> */}
```
Comment (b0b3rt) [inline packages/lesswrong/components/bookmarks/BookmarksPage.tsx:70]: ?

File: packages/lesswrong/components/common/LWHomePosts.tsx [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        setCookie(RECOMBEE_SETTINGS_COOKIE, JSON.stringify([[getDefaultTab(currentUser), defaultRecombeeConfig]]), { path: '/' });
+      }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, []);
+  
+    return {
+      selectedScenario, updateSelectedScenario,
+      scenarioConfig, updateScenarioConfig
+    };
+  }
```
Comment (b0b3rt) [inline packages/lesswrong/components/common/LWHomePosts.tsx:163]: This function should not be inlined in the component

File: packages/lesswrong/components/common/LWHomePosts.tsx [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+          <TabPicker 
+            sortedTabs={enabledAlgorithms} 
+            defaultTab={selectedTab} 
+            onTabSelectionUpdate={handleSwitchTab}
+            showDescriptionOnHover
+          />
+          {showSettingsButton && settingsButton}
+        </div>
+        {settings}
+        {isFriendlyUI && <StickiedPosts />}
```
Comment (b0b3rt) [inline packages/lesswrong/components/common/LWHomePosts.tsx:297]: I think we can delete this line

File: packages/lesswrong/components/common/LWHomePosts.tsx [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+        {/* TODO: reenable, disabled for testing to see how often duplication happens */}
+        <HideRepeatedPostsProvider>
+          <AnalyticsContext listContext={"latestPosts"}>
+            {/* Allow hiding posts from the front page*/}
+            <AllowHidingFrontPagePostsContext.Provider value={true}>
+
+              {/* LATEST POSTS (Hacker News Algorithm) */}
```
Comment (b0b3rt) [inline packages/lesswrong/components/common/LWHomePosts.tsx:304]: Surely, if we're always wrapping whatever posts we're showing in `<AnalyticsContext feedType={selectedTab}>`, we should not repeat that for every case?

File: packages/lesswrong/components/common/LWHomePosts.tsx [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    const testingFeeds = postFeedsTestingSetting.get().map(feed => ({ name: feed.name, label: feed.label, description: feed.description, disabled: feed.disabled }));
+    availableTabs.push(...testingFeeds);
+  }
+
+  const enabledTabs = availableTabs
+    .filter(feed => !feed.disabled
+      && !(feed.name === 'lesswrong-bookmarks' && (countBookmarks ?? 0) < 1)
+      && !(feed.name === 'lesswrong-continue-reading' && continueReading?.length < 1)
+    )
+  const [ selectedTab, setSelectedTab ] = useState(getDefaultTab(currentUser, enabledTabs));
```
Comment (b0b3rt) [inline packages/lesswrong/components/common/LWHomePosts.tsx:150]: nit: no spaces inside array destructuring

File: packages/lesswrong/components/common/LWHomePosts.tsx [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+                onTabSelectionUpdate={handleSwitchTab}
+                showDescriptionOnHover
+              />
+            </div>
+            {showSettingsButton && settingsButton}
+          </>}
+        </div>
+        {settings}
+        {isFriendlyUI && <StickiedPosts />}
+        {/* TODO: reenable, disabled for testing to see how often duplication happens */}
```
Comment (b0b3rt) [inline packages/lesswrong/components/common/LWHomePosts.tsx:307]: delete outdated todo

File: packages/lesswrong/lib/collections/users/schema.ts [hunk 7]
```diff
@@ -941,6 +941,15 @@ const schema: SchemaType<"Users"> = {
     canCreate: 'guests',
     hidden: true,
   },
+  frontpageSelectedTab: {
+    type: String,
+    optional: true,
+    nullable: true,
+    canRead: userOwns,
+    canUpdate: [userOwns, 'sunshineRegiment', 'admins'],
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/users/schema.ts:949]: I think this probably shouldn't include `sunshineRegiment`; it's not moderation-related

File: packages/lesswrong/lib/collections/users/schema.ts [hunk 8]
```diff
@@ -941,6 +941,15 @@ const schema: SchemaType<"Users"> = {
     canCreate: 'guests',
     hidden: true,
   },
+  frontpageSelectedTab: {
+    type: String,
+    optional: true,
+    nullable: true,
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/users/schema.ts:947]: Hm... my guess is that this should not be nullable, and should have a default value of whatever the name of the latest posts tab is.

File: packages/lesswrong/lib/collections/users/schema.ts (hunk 7 above)
Comment (b0b3rt) [reply packages/lesswrong/lib/collections/users/schema.ts:949]: Well... we'd want to not include `lesswrong` in the tab name, in that case.

File: packages/lesswrong/lib/collections/users/schema.ts (hunk 8 above)
Comment (b0b3rt) [reply packages/lesswrong/lib/collections/users/schema.ts:947]: (In fact I think we probably shouldn't include `lesswrong` in any of the tab names; we can just omit it, or if we want to specifically indicate that they aren't recombee tabs then we can say `forum-` instead.)

**Review by oetherington — COMMENTED** [review]
No objections, apart from one missing forum gate

**Pull request merged**